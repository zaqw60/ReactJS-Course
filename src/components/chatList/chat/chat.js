import Divider from "@material-ui/core/Divider"
import ListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"

export function Chat({ title, selected, handleListItemClick, lastMessage }) {
  return (
    <>
      <ListItem
        className="item"
        button={true}
        selected={selected}
        onClick={handleListItemClick}
      >
        <ListItemText className="text" primary={title} />
        {lastMessage && (
          <ListItemText
            className="text"
            primary={`${lastMessage.author}: ${lastMessage.message}`}
          />
        )}
        <ListItemText className="text" primary="12.30" />
      </ListItem>
      <Divider />
    </>
  )
}
