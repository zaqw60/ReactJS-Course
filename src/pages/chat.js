import { useEffect } from "react"
import { useDispatch } from "react-redux"
import { Switch, Route, useHistory } from "react-router-dom"
import { Layout, ChatList, MessageList } from "../components"
import { getConversationsFB } from "../store/conversations"
import { getMessagesFB } from "../store/messages"

export function Chat() {
  const { push } = useHistory()
  useEffect(() => {
    const listenExistChat = ({ code }) => {
      if (code === "Escape") {
        push("/chat")
      }
    }

    document.addEventListener("keydown", listenExistChat)

    return () => {
      document.removeEventListener("keydown", listenExistChat)
    }
  }, [push])

  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(getConversationsFB())
    dispatch(getMessagesFB())
  }, [dispatch])

  return (
    <Switch>
      <Route path={["/chat/:roomId", "/chat"]}>
        <Layout chats={<ChatList />}>
          <Route path="/chat/:roomId">
            <MessageList />
          </Route>
          <Route exact={true} path="/chat">
            <h1>выберите или добавьте чат </h1>
          </Route>
        </Layout>
      </Route>
      <Route path="*">
        <h1>такого чата нет</h1>
      </Route>
    </Switch>
  )
}
