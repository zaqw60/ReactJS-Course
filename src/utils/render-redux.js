import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { BrowserRouter } from "react-router-dom"
import { createStore } from "redux"
import { reducer } from "../store"

export const renderRedux = (component, { initialState }) => {
  const store = createStore(reducer, initialState)

  return {
    store,
    ...render(
      <Provider store={store}>
        <BrowserRouter>{component}</BrowserRouter>
      </Provider>,
    ),
  }
}
