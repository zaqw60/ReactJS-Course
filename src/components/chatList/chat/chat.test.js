import { getByTestId } from "@testing-library/react"
import { renderRedux } from "../../../utils/render-redux"
import { Chat } from "./chat"

let state = null

beforeEach(() => {
  state = {
    messages: {
      messages: { room1: [{ author: "User", message: "Test" }] },
    },
  }
})

describe("chat component", () => {
  it("test render chat", () => {
    const { container } = renderRedux(<Chat title="room1" />, {
      initialState: state,
    })

    const nodes = [...container.querySelectorAll(".text")]

    expect(nodes[0]).toHaveTextContent("room1")
    expect(nodes[1]).toHaveTextContent("User: Test")
  })

  it(" render chat with selected prop", () => {
    const { container } = renderRedux(<Chat title="room1" selected={true} />, {
      initialState: state,
    })

    expect(container.querySelector(".item")).toHaveClass("Mui-selected")
  })

  it("snapshot", () => {
    const { container } = renderRedux(<Chat title="room1" />, {
      initialState: state,
    })
    expect(container).toMatchSnapshot()
  })
})
