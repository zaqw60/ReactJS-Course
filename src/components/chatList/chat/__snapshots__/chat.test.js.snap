// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat component snapshot 1`] = `
<div>
  <a
    href="/chat/room1"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root item MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemText-root text"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          room1
        </span>
      </div>
      <div
        class="MuiListItemText-root text"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          User: Test
        </span>
      </div>
      <div
        class="MuiListItemText-root text"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          12.30
        </span>
      </div>
      <button
        aria-label="delete"
        class="MuiButtonBase-root MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <hr
      class="MuiDivider-root"
    />
  </a>
</div>
`;
